euphonic/fast_adaptive_broadening.py

```diff
@@ -0,0 +1,166 @@
+"""
+Functions for fast adaptive broadening of density of states spectra
+"""
+from typing import Optional
+from scipy.optimize import curve_fit
+from scipy.signal import convolve
+import numpy as np
+from euphonic import Quantity
+
+
+def fast_broaden(dos_bins: np.ndarray,
+                 freqs: np.ndarray,
+                 mode_widths: Quantity,
+                 weights: np.ndarray,
+                 mode_weights: np.ndarray,
+                 adaptive_error: float) -> np.ndarray:
+    """
+    Uses a fast, approximate method to adaptively broaden a density
+    of states spectrum
+
+    Parameters
+    ----------
+    dos_bins
+        Shape (n_e_bins + 1,) float ndarray. The energy bin edges to use for
+        calculating the DOS
+    freqs
+        Shape (n_qpts, n_modes) float ndarray. Frequencies per q-point
+        and mode.
+    mode_widths
+        Shape (n_qpts, n_modes) float Quantity in energy units. The broadening
+        width for each mode at each q-point
+    weights
+        Shape (n_qpts,) float ndarray. The weight for each q-point.
+    mode_weights
+        Shape (n_qpts, n_modes) float ndarray. The weight of each mode at
+        each q-point.
+    adaptive_error
+        Scalar float. Acceptable error for gaussian approximations, defined
+        as the absolute difference between the areas of the true and
+        approximate gaussians.
+
+    Returns
+    -------
+    dos
+        Float ndarray of shape (dos_bins - 1,) containing density of states
+        ydata
+    """
+    n_modes = freqs.shape[-1]
+    freqs = np.ravel(freqs)
+    mode_weights = np.ravel(mode_weights)
+    mode_widths = np.ravel(mode_widths).to('hartree').magnitude
+    n_weights = np.repeat(weights, n_modes)
+
+    # set default error value if none specified
+    if adaptive_error is None:
+        adaptive_error = 0.01
+
+    # determine spacing value for mode_width samples given desired error level
+    spacing = 656.1*adaptive_error**3 - 131.8*adaptive_error**2 +\
+                15.98*adaptive_error + 1.0803
+
+    bin_width = dos_bins[1]-dos_bins[0]
+    mode_widths = np.maximum(mode_widths, bin_width/2)
+
+    n_kernels = int(
+        np.ceil(np.log(max(mode_widths)/min(mode_widths))/np.log(spacing)))
+    mode_width_samples = spacing**np.arange(n_kernels+1)*min(mode_widths)
+    freq_range = 3*max(mode_widths)
+    kernel_npts_oneside = np.ceil(freq_range/bin_width)
+    kernels = gaussian(xvals=np.arange(-kernel_npts_oneside,
+                       kernel_npts_oneside+1, 1)*bin_width,
+                       sigma=mode_width_samples[:, np.newaxis],
+                       centre=0)*bin_width
```

Comment:
Looking at all these "kernel_npts_oneside" reminded me that we do have the option to evaluate the kernels on one side and mirror them to make the full symmetric function. A quick play around in IPython suggests that
```
scipy.stats.norm.pdf(x, scale=sigma)
```
is _slightly_ faster than
```
half_pdf  scipy.stats.norm.pdf(half_x, scale=sigma)
pdf = np.concatenate([np.flip(half_pdf[:, 1:], axis=1), half_pdf], axis=1)
```
where x is 1001 points, half_x is 501 points and sigma is a column vector of 20 widths. But if we bump the numbers a bit to 4001 points and 50 sigma values, the mirror implementation is nearly twice as fast. 

I think what is happening is that `norm.pdf` has better-than-linear scaling for small numbers of points, as the meaningful work grows larger relative to overheads. Once we hit the actual linear-scaling regime it becomes possible for the factor-2 saving to outweight the overheads.

This transition point is uncomfortably close to the kind of numbers we are looking at using in practice. I guess for now we can stick to a simple implementation and then if kernel generation is a real-world bottleneck _then_ consider introducing this mirror technique.